File: emisor/algoritmos/fletcher.py
"""Checksum Fletcher configurable para palabras de 8, 16 o 32 bits."""

from algoritmos.contrato import EstadoVerificacion, ResultadoVerificacion, TramaCodificada


class Fletcher:
    nombre = "fletcher"

    def __init__(self, tamano_bloque: int = 16):
        if tamano_bloque not in (8, 16, 32):
            raise ValueError("tamano_bloque debe ser 8, 16 o 32")
        self.tamano_bloque = tamano_bloque

    @staticmethod
    def _checksum(bits: str, tamano: int) -> int:
        modulo = 2**tamano - 1
        suma1 = suma2 = 0
        for i in range(0, len(bits), tamano):
            suma1 = (suma1 + int(bits[i : i + tamano], 2)) % modulo
            suma2 = (suma2 + suma1) % modulo
        return (suma2 << tamano) | suma1

    def calcular(self, bits: str) -> TramaCodificada:
        relleno = (-len(bits)) % self.tamano_bloque
        datos_rellenos = bits + "0" * relleno
        checksum = self._checksum(datos_rellenos, self.tamano_bloque)
        trama = datos_rellenos + format(checksum, f"0{2 * self.tamano_bloque}b")
        parametros = {"tamano_bloque": self.tamano_bloque, "bits_datos": len(bits)}
        return TramaCodificada(trama, parametros)

    def verificar(self, trama: str, parametros: dict) -> ResultadoVerificacion:
        tamano = parametros["tamano_bloque"]
        bits_datos = parametros["bits_datos"]
        longitud_rellena = bits_datos + (-bits_datos) % tamano
        datos = trama[:longitud_rellena]
        recibido = trama[longitud_rellena:]
        esperado = format(self._checksum(datos, tamano), f"0{2 * tamano}b")
        if recibido == esperado:
            return ResultadoVerificacion(EstadoVerificacion.SIN_ERROR, datos[:bits_datos], "checksum coincide")
        return ResultadoVerificacion(EstadoVerificacion.ERROR_NO_CORREGIBLE, None, "checksum Fletcher no coincide")

File: receptor/algoritmos/fletcher.js
/** Checksum Fletcher configurable para palabras de 8, 16 o 32 bits. */
import { AlgoritmoIntegridad, EstadoVerificacion } from "./contrato.js";

export class Fletcher extends AlgoritmoIntegridad {
  constructor(tamanoBloque = 16) {
    super();
    if (![8, 16, 32].includes(tamanoBloque)) throw new Error("tamano_bloque debe ser 8, 16 o 32");
    this.tamanoBloque = tamanoBloque;
  }

  static checksum(bits, tamano) {
    const modulo = (1n << BigInt(tamano)) - 1n;
    let suma1 = 0n;
    let suma2 = 0n;
    for (let i = 0; i < bits.length; i += tamano) {
      suma1 = (suma1 + BigInt(`0b${bits.slice(i, i + tamano)}`)) % modulo;
      suma2 = (suma2 + suma1) % modulo;
    }
    return (suma2 << BigInt(tamano)) | suma1;
  }

  calcular(bits) {
    const relleno = (this.tamanoBloque - (bits.length % this.tamanoBloque)) % this.tamanoBloque;
    const datos = bits + "0".repeat(relleno);
    const checksum = Fletcher.checksum(datos, this.tamanoBloque).toString(2).padStart(2 * this.tamanoBloque, "0");
    return { bits: datos + checksum, parametros: { tamano_bloque: this.tamanoBloque, bits_datos: bits.length } };
  }

  verificar(trama, parametros) {
    const tamano = parametros.tamano_bloque;
    const longitud = parametros.bits_datos + (tamano - (parametros.bits_datos % tamano)) % tamano;
    const datos = trama.slice(0, longitud);
    const esperado = Fletcher.checksum(datos, tamano).toString(2).padStart(2 * tamano, "0");
    if (trama.slice(longitud) === esperado) {
      return { estado: EstadoVerificacion.SIN_ERROR, bits: datos.slice(0, parametros.bits_datos), detalle: "checksum coincide" };
    }
    return { estado: EstadoVerificacion.ERROR_NO_CORREGIBLE, bits: null, detalle: "checksum Fletcher no coincide" };
  }
}
